driver-core/src/main/com/mongodb/Jep395RecordCodecProvider.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb;
+
+import com.mongodb.internal.VisibleForTesting;
+import com.mongodb.lang.Nullable;
+import org.bson.codecs.Codec;
+import org.bson.codecs.configuration.CodecProvider;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.codecs.record.RecordCodecProvider;
+
+import static com.mongodb.internal.VisibleForTesting.AccessModifier.PRIVATE;
+
+
+/**
+ * A CodecProvider for Java Records.
+ *
+ * <p>Requires java.lang.Record support - eg Java 17 or greater.</p>
+ */
+public class Jep395RecordCodecProvider implements CodecProvider {
```

Comment:
Yes, we would need to have version specific substitutions to remove that class on java 11. Also, while  there is work in progress, records are not native friendly yet.

Reply:
@christophstrobl, @mp911de, this may be a Spring Boot issue as well.

It's not clear to me what the way forward is for this and similar issues.  Do all of these substitutions (the one for DNS provider as well) have to done separately for every framework that supports native images?

Reply:
Yeah, BUT it's precisely what we want to avoid doing. We want to know almost everything at build time to optimize and initialize during the build and not during the runtime. I know it can be seen as a different way to execute Java (it is), but the benefits are quite high.

Reply:
Thanks for the clarification @cescoffier :+1: